### PR TITLE
feat: JSON-RPC external analyzer protocol for multi-language support (Phase 2 of #95)

### PR DESCRIPTION
**Pull request by em-redhat**
## Summary

Phase 2 of Issue #95. Adds a JSON-RPC 2.0 over stdin/stdout protocol that lets gaze spawn external language analyzer binaries and produce CRAP scores, GazeCRAP, quadrants, and fix strategies for any language.

Builds on Phase 1 (PR #165) which extracted four provider interfaces from the scoring pipeline.

### New: JSON-RPC Protocol Client (`internal/protocol/`)

- `Client` struct managing subprocess lifecycle (spawn, stdin/stdout communication, shutdown)
- `Call(ctx, method, params)` with `context.Context` timeout support
- 8 protocol methods: `initialize`, `discover`, `analyze`, `complexity`, `coverage`, `test_mapping`, `classify_signals`, `shutdown`
- Capability negotiation via `initialize` response
- Fake analyzer binary at `testdata/fake_analyzer/` for testing

### New: External Provider Adapters (`internal/adapter/`)

- `ExternalComplexityProvider` — translates `complexity` responses to `[]FunctionComplexity`
- `ExternalLineCoverageProvider` — translates `coverage` responses to `[]FuncCoverage`
- `ExternalSideEffectAnalyzer` — translates `analyze` responses to `[]AnalysisResult` with classification
- `ExternalContractCoverageProvider` — uses `test_mapping` + `ComputeContractCoverage` for GazeCRAP
- `Session` — manages full protocol lifecycle, returns constructed adapters
- `Discover` — three-tier analyzer discovery (CLI flag → `.gaze.yaml` → PATH convention)

### New: CLI Flags

- `--analyzer <binary>` on `crap`, `quality`, `report` commands
- `--language <lang>` for config/PATH-based discovery
- When `--analyzer` is set, external adapters are used; when not set, behavior is identical to today

### New: Configuration

```yaml
# .gaze.yaml
analyzers:
  python:
    command: snake-eyes
    args: ["--stdio"]
```

### New: Protocol Specification

`docs/protocol.md` — full reference for external analyzer authors

## Backward Compatibility

- Zero changes when `--analyzer` is not specified
- All 16 packages pass tests
- E2E self-check path unchanged
- No new required dependencies

## Testing

- 16 protocol client tests (session, errors, timeout, crash recovery)
- 11 adapter tests (translation, session lifecycle, capability negotiation)
- 6 discovery tests (CLI flag, config, PATH, edge cases)
- 3 CLI integration tests with fake analyzer
- All existing tests pass (zero regression)

## OpenSpec Artifacts

`openspec/changes/external-analyzer-protocol/` — proposal, design, specs, tasks (27/27 complete)

Ref: #95

### CODE REVIEW

**Review by yvonnedevlinrh — DISMISSED** [review]
LGTM — well-architected feature that cleanly adds multi-language support via JSON-RPC 2.0 protocol. The layering (protocol → adapter → provider interface) maintains clean separation of concerns, and the external path is properly isolated from the Go-native path. 36+ tests cover protocol lifecycle, adapter translation, discovery tiers, and CLI integration. Fully backward-compatible — no behavior change when --analyzer is not set.

One MEDIUM observation: buildContractLookup in internal/adapter/contract.go constructs ContractCoverageInfo with only Percentage, leaving Reason, MinConfidence, and MaxConfidence as zero values. The Go-native path populates these with diagnostic data (e.g., "all effects ambiguous, confidence 78-79"). External analyzer users will see contract coverage percentages but no explanation when coverage is 0%. Worth tracking for a follow-up to bring external output parity with Go-native output.

Three LOW observations (non-blocking):
- Error.Error() in internal/protocol/types.go has a dead if branch — Data is checked but never included in the output (both branches return e.Message)
- language parameter in internal/adapter/discovery.go is concatenated into exec.LookPath without validating it contains only [a-z0-9-] characters (theoretical, not exploitable — input comes from CLI flag or config)
- Trust model for config-sourced binary execution (analyzers.<lang>.command in .gaze.yaml) is not documented — same trust model as Makefile/goreleaser, but a comment would improve transparency

APPROVE

Reviewed-by: yvonnedevlinrh
Assisted-by: OpenCode (claude-opus-4-6)

File: cmd/gaze/main.go [hunk 1]
```diff
@@ -1197,6 +1328,10 @@ Requires the target package to have existing test files.`,
 		"AI backend for assertion mapping fallback: claude, gemini, ollama, or opencode")
 	cmd.Flags().StringVar(&aiMapperModel, "ai-mapper-model", "",
 		"model name for AI mapper (required for ollama)")
+	cmd.Flags().StringVar(&analyzerFlag, "analyzer", "",
```
Comment (jflowers) [inline cmd/gaze/main.go:1447]: **[LOW] Consider hiding `--analyzer` on quality until supported.**

The `--analyzer` flag is registered on the quality command but immediately errors when used (deferred per D12). Consider adding `_ = cmd.Flags().MarkHidden("analyzer")` and `_ = cmd.Flags().MarkHidden("language")` after registration to avoid confusing users browsing `--help`, while keeping the code wired for future implementation.

Non-blocking — the error message is clear and informative.

**Review by jflowers — DISMISSED** [review]
**APPROVE** — Well-architected feature that cleanly adds multi-language support via JSON-RPC 2.0 protocol.

The layering (protocol → adapter → provider interface) maintains clean separation of concerns, and the external path is properly isolated from the Go-native path. 36+ tests cover protocol lifecycle, adapter translation, discovery tiers, and CLI integration. Fully backward-compatible — no behavior change when `--analyzer` is not set.

Prior reviewer's four findings (ContractCoverageInfo parity, Error.Error() dead branch, language validation, trust model docs) have all been addressed in commit 131cd1f.

One LOW inline comment on the quality command's `--analyzer` flag visibility.

_This review was generated by /review-pr (AI-assisted)._

File: cmd/gaze/main.go [hunk 2]
```diff
@@ -1318,6 +1449,10 @@ Packages without test files are skipped with a warning.`,
...
```
Comment (jflowers) [inline cmd/gaze/main.go:1447]: **[LOW] Consider hiding `--analyzer` on quality until supported.**

The `--analyzer` and `--language` flags are registered on the quality command but immediately error when used (deferred per D12). Consider adding `_ = cmd.Flags().MarkHidden("analyzer")` and `_ = cmd.Flags().MarkHidden("language")` after registration to avoid confusing users browsing `--help`.

Non-blocking — the error message is clear and informative. Previously raised in the dismissed review.

**Review by jflowers — DISMISSED** [review]
**APPROVE** — Post-merge-resolution re-review.

The merge conflict resolution (integrating `analyze-multi-package` from upstream) is clean and correct. The `--analyzer` rejection guard in `runQuality` fires before the multi-package logic. The test was correctly updated to use the new `patterns` field. All local checks pass (build, vet, lint, 16 packages with `-race -count=1 -short`).

The core feature (JSON-RPC external analyzer protocol) was reviewed in the previous cycle. Prior reviewer's four findings were all addressed. One LOW finding carried forward (see inline comment).

_This review was generated by /review-pr (AI-assisted)._

File: internal/protocol/client.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+		return nil // already closed
+	}
+
+	// Send shutdown request with a short timeout. Best-effort —
+	// if the analyzer is already dead, the write will fail and
+	// we proceed to cleanup.
+	shutdownCtx, cancel := context.WithTimeout(context.Background(), shutdownTimeout)
+	defer cancel()
+
+	_, _ = c.Call(shutdownCtx, MethodShutdown, nil)
```
Comment (gxmiranda) [inline internal/protocol/client.go]: **[HIGH] `Close()` never sends the shutdown request — state ordering bug.**

`Close()` calls `c.closed.Swap(true)` on line 218, then calls `c.Call()` here on line 228. But `Call()` checks `c.closed.Load()` on line 131 and returns `fmt.Errorf("protocol client is closed")` if true.

Since `Close()` sets `closed=true` *before* calling `Call()`, the shutdown request is silently dropped every time. The `_, _` assignment hides a guaranteed error. The subprocess exits because stdin is closed on line 231, but it never receives the `shutdown` JSON-RPC message.

Analyzers that rely on the shutdown signal for cleanup (flushing logs, releasing resources, writing coverage data) will not execute their cleanup logic.

**Fix options**:
1. Move `c.closed.Swap(true)` after the `Call()` and `stdin.Close()`
2. Add an internal `sendRaw()` method that bypasses the closed check, used only by `Close()`
3. Use a separate `closing` state that `Call()` allows for the shutdown method only

Option 2 is cleanest — `Close()` needs to send one final message without the guard that protects normal callers from using a closed client.

File: cmd/gaze/main.go (hunk 2 above)
Comment (gxmiranda) [inline cmd/gaze/main.go:1447]: **[MEDIUM] Consider hiding `--analyzer` and `--language` on quality until supported.**

These flags are registered but immediately error when used (deferred per D12). They appear in `gaze quality --help` output, which may confuse users browsing available options.

```suggestion
	cmd.Flags().StringVar(&analyzerFlag, "analyzer", "",
		"external analyzer binary (e.g., snake-eyes)")
	cmd.Flags().StringVar(&languageFlag, "language", "",
		"target language for analyzer discovery (e.g., python)")
	_ = cmd.Flags().MarkHidden("analyzer")
	_ = cmd.Flags().MarkHidden("language")
```

Non-blocking — the error message is clear and informative. Previously raised by @jflowers in both prior review cycles.

**Review by gxmiranda — COMMENTED** [review]
**COMMENT** — Well-architected feature that cleanly adds multi-language support via JSON-RPC 2.0 protocol.

The layering (protocol → adapter → provider interface) maintains clean separation of concerns, and the external path is properly isolated from the Go-native path. 36+ tests cover protocol lifecycle, adapter translation, discovery tiers, and CLI integration. Fully backward-compatible — no behavior change when `--analyzer` is not set.

All 4 constitution principles verified (Accuracy, Minimal Assumptions, Actionable Output, Testability) — PASS on all. Prior reviewer findings (ContractCoverageInfo parity, Error.Error() dead branch, language validation, trust model docs) were all addressed.

**Findings**: 1 HIGH, 2 MEDIUM, 2 LOW.

- **[HIGH]** `Close()` never sends shutdown request due to state ordering bug (see inline comment on `client.go:228`)
- **[MEDIUM]** `--analyzer` flag visible on `quality --help` but immediately errors (see inline comment on `main.go:1452`; previously raised by @jflowers)
- **[MEDIUM]** `Discover()` Tier 1 skips language validation for CLI flag path — low-risk since value is user-controlled, and goes through `exec.LookPath` in `NewClient()`. Trust model GoDoc on `Discover()` is thorough.
- **[LOW]** `SideEffectAnalyzer` not wired through `crap.Options` — consistent with D12 deferral, noted for Phase 3
- **[LOW]** Missing trailing newline in `proposal.md` and `spec.md`

_This review was generated by /review-pr (AI-assisted)._

File: internal/protocol/client.go (hunk 3 above)
Comment (em-redhat) [reply internal/protocol/client.go]: Addressed in `936fa98`: Added `sendShutdown()` method that bypasses the `closed` state guard (option 2 from the review). `Close()` now sends the shutdown JSON-RPC message before closing stdin, ensuring analyzers receive the shutdown signal for cleanup.

File: cmd/gaze/main.go (hunk 2 above)
Comment (em-redhat) [reply cmd/gaze/main.go:1447]: Addressed in `936fa98`: Added `MarkHidden("analyzer")` and `MarkHidden("language")` on the quality command. Flags remain wired for future D12 implementation but no longer appear in `--help`.

Comment (em-redhat) [reply cmd/gaze/main.go:1447]: Addressed in `936fa98` alongside @gxmiranda's equivalent finding. Both flags now hidden via `MarkHidden`.

**Review by gxmiranda — APPROVED** [review]
**APPROVE** — Well-architected feature that cleanly adds multi-language support via JSON-RPC 2.0 protocol.

The layering (protocol → adapter → provider interface) maintains clean separation of concerns, and the external path is properly isolated from the Go-native path. 36+ tests cover protocol lifecycle, adapter translation, discovery tiers, and CLI integration. Fully backward-compatible — no behavior change when `--analyzer` is not set.

All prior reviewer findings (Close() shutdown bug, ContractCoverageInfo parity, Error.Error() dead branch, language validation, trust model docs, hidden quality flags) have been addressed in subsequent commits.

Constitution compliance verified: Accuracy (scoring engine unchanged), Minimal Assumptions (protocol is language-agnostic), Actionable Output (same formatting/fix strategies), Testability (fake binary + DI throughout).

3 LOW observations (non-blocking): missing trailing newlines in spec artifacts, SideEffectAnalyzer not wired to crap.Options (consistent with D12 deferral), verbose `containsAll` helper in tests.

_This review was generated by /review-pr (AI-assisted)._

**Review by marcusburghardt — APPROVED** [review]
Nice work @em-redhat
